tests/test_spt_set.py

```diff
@@ -0,0 +1,124 @@
+    assert contract.getFirstEmptySlot() == 0
+    assert contract.getDepth() == 1
+
+def test_remove_last(accounts):
+    contract = PublicSmtSetSha.deploy(10, {"from": accounts[0]})
+    contract.addToNextEmpty(b"apple")
+    contract.addToNextEmpty(b"fish")
+    contract.addToNextEmpty(b"banana")
+    contract.addToNextEmpty(b"ice")
+    assert contract.getRoot() == get_root_4([b"apple", b"fish", b"banana", b"ice"])
+    contract.removeAndRebase(3)
+    assert contract.getRoot() == get_root_4([b"apple", b"fish", b"banana", b""])
+
+def test_remove_first(accounts):
+    contract = PublicSmtSetSha.deploy(10, {"from": accounts[0]})
+    contract.addToNextEmpty(b"apple")
+    contract.addToNextEmpty(b"fish")
+    contract.addToNextEmpty(b"banana")
+    contract.addToNextEmpty(b"ice")
+    assert contract.getRoot() == get_root_4([b"apple", b"fish", b"banana", b"ice"])
+    contract.removeAndRebase(0)
+    assert contract.getRoot() == get_root_4([b"ice", b"fish", b"banana", b""])
+
+def test_remove_second(accounts):
+    contract = PublicSmtSetSha.deploy(10, {"from": accounts[0]})
+    contract.addToNextEmpty(b"apple")
+    contract.addToNextEmpty(b"fish")
+    contract.addToNextEmpty(b"banana")
+    contract.addToNextEmpty(b"ice")
+    assert contract.getRoot() == get_root_4([b"apple", b"fish", b"banana", b"ice"])
+    contract.removeAndRebase(1)
+    assert contract.getRoot() == get_root_4([b"apple", b"ice", b"banana", b""])
+
+def test_limit(accounts):
+    contract = PublicSmtSetSha.deploy(4, {"from": accounts[0]})
+
+    for i in range(0, 2**4):
+        bs = b"just" + (b"bla" * i)
+        contract.addToNextEmpty(bs)
+        assert contract.getFirstEmptySlot() == i + 1
+
+    reverted = False
+    try:
+        tx = contract.addToNextEmpty(b"try")
+    except Exception as e:
+        reverted = True
+        assert "Index out of range" in e.message
+
+    assert reverted
+
+def test_sizes(accounts):
+    contract = PublicSmtSetSha.deploy(8, {"from": accounts[0]})
+
+    for exp in range(1, 8):
+
+        for i in range(0, 2**exp):
+            assert contract.getFirstEmptySlot() == i
+            if i == 0 or i == 1:
+                depth = 1
+            else:
+                depth = math.ceil(math.log2(i))
+            bs = b"just" + (b"bla" * i)
+            contract.addToNextEmpty(bs)
+            assert contract.getFirstEmptySlot() == i + 1
+            if i + 1 == 0 or i + 1 == 1:
+                depth = 1
+            else:
+                depth = math.ceil(math.log2(i + 1))
+            assert contract.getDepth() == depth
+
+        for i in range(0, 2**exp):
+            assert contract.getFirstEmptySlot() == 2**exp - i
+            contract.removeAndRebase(0)
+            q = 2**exp - i - 1
+            assert contract.getFirstEmptySlot() == q 
+            if q == 0 or q == 1:
+                depth = 1
+            else:
+                depth = math.ceil(math.log2(q))
+            assert contract.getDepth() == depth
```

Comment:
nit: no newline at the end of file

Reply:
done https://github.com/andreysobol/sparse_merkle_sol/pull/7/commits/9597dfe3b8bccf78abc1090da2ed863852145752